### PR TITLE
Add first version

### PR DESCRIPTION
**Pull request by martin-ejdestig**

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,31 @@
+project('template-dbus-service', 'cpp',
+    version : '0.1.0',
+    meson_version : '>=0.45',
+    default_options : [
+        'cpp_std=c++17',
```
Comment (JEderonn) [inline meson.build:5]: Is the template using C++17 features?

Comment (martin-ejdestig) [reply meson.build:5]: Yes.

File: src/daemon/dbus_service.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            return;
+
+        connection_id_ = Gio::DBus::own_name(Gio::DBus::BUS_TYPE_SYSTEM,
+                                             Common::DBus::TEMPLATE_SERVICE_NAME,
+                                             sigc::mem_fun(*this, &DBusService::bus_acquired),
+                                             sigc::mem_fun(*this, &DBusService::name_acquired),
+                                             sigc::mem_fun(*this, &DBusService::name_lost));
+    }
+
+    void DBusService::unown_name()
```
Comment (JEderonn) [inline src/daemon/dbus_service.cpp:39]: Ha ha, fun naming: 'unown' :). A bit difficult to read/understand, consider rename method to not_own_name()  or release_name()?

Comment (martin-ejdestig) [reply src/daemon/dbus_service.cpp:39]: Hum, I think it should be kept. It makes sense if you look at the GDBus API (just a "wrapper" around Gio::DBus::unown_name()).

But I can change it if you really want to.

Comment (martin-ejdestig) [reply src/daemon/dbus_service.cpp:39]: I think I prefer `release_name()` over `not_own_name()`.

File: src/daemon/dbus_service.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    public:
+        explicit DBusService(const Glib::RefPtr<Glib::MainLoop> &main_loop);
+        ~DBusService() final;
+
+        DBusService(const DBusService &other) = delete;
+        DBusService(DBusService &&other) = delete;
+        DBusService &operator=(const DBusService &other) = delete;
+        DBusService &operator=(DBusService &&other) = delete;
+
+        void own_name();
```
Comment (TSAnsari) [inline src/daemon/dbus_service.h:30]: Then this may be named acquire_name() ?

Comment (martin-ejdestig) [reply src/daemon/dbus_service.h:30]: I would rather keep `own` since it is "D-Bus terminology". But I don't know. Does it matter that much? It is not like `own` does not go together as the opposite for `release`.

Comment (martin-ejdestig) [reply src/daemon/dbus_service.h:30]: And vice versa.

Comment (martin-ejdestig) [reply src/daemon/dbus_service.h:30]: But, hum... they mix it in GDBus. OK, whatever, I'll change it. :)

Comment (TSAnsari) [reply src/daemon/dbus_service.h]: One own's something after acquiring it. Own itself is not the process of acquiring. Is unown also dbus term?

File: README.md [hunk 4]
```diff
@@ -1,2 +1,12 @@
-# template-dbus-service
-Template for creating a D-Bus service component
+# Template D-Bus service
+
+Template for creating a D-Bus service component.
+
```
Comment (sashko) [inline README.md:4]: so many empty lines....

File: src/daemon/dbus_service.cpp (hunk 2 above)
Comment (JEderonn) [reply src/daemon/dbus_service.cpp:39]: Perhaps it's bad to use another terminology. I was reading the code too fast. I thought unown was a new term but apparently it wasn't. So perhaps you should keep own/unown, sorry Martin :(.

File: meson.build (hunk 1 above)
Comment (martin-ejdestig) [reply meson.build:5]: E.g. std::optional, nested namespace definitions, constexpr variables are implicitly inlined... and... perhaps I missed something. Nothing big, but nice to have. Was not that much new in C++17 compared to C++14.

File: src/daemon/dbus_service.h (hunk 3 above)
Comment (martin-ejdestig) [reply src/daemon/dbus_service.h:30]: @TSAnsari Did you remove a comment here? Anyway, they talk more about "owning a name" in the D-Bus repository than "acquiring a name".

Will probably revert both fixup commits related to this on Monday. It does not matter that much but probably just confusing to call it something else when the Glib::GDBus functions that are invoked are called `own_name()` and `unown_name()`.

File: meson.build (hunk 1 above)
Comment (martin-ejdestig) [reply meson.build:5]: To clarify a bit more if your worries are about support. It is all features GCC and Clang have supported for a very long time (at least ~2 years in a stable form in a stable release). You can check https://en.cppreference.com/w/cpp/compiler_support for an overview but have to look at release notes for GCC and Clang for more detailed information.

File: README.md (hunk 4 above)
Comment (martin-ejdestig) [reply README.md:4]: Yes, put there intentionally to make it more readable.

**Comment by martin-ejdestig** [comment]
By the way, glibmm documentation on the Web is working now again:
- https://developer.gnome.org/glibmm/stable/

And as mentioned before, some times reading the original C documentation after (straight forward) mapping of method/class from C++ to C helps/is better:
- https://developer.gnome.org/glib/stable/
- https://developer.gnome.org/gio/stable/

`devhelp`, that I showed briefly last week, is a nice application for browsing and searching GNOME/GLib/glibmm related documentation locally. Available in all major Linux dists.

Also, Meson documentation can be found here if you have not already figured it out by yourself:
- https://mesonbuild.com/Manual.html
- https://mesonbuild.com/Reference-manual.html

Search bar can be found in the upper right corner.

File: README.md (hunk 4 above)
Comment (martin-ejdestig) [reply README.md:4]: Use the space, @sashko. But, if feelings about it you have, space remove I can. Matters to me much, it does not.

**Comment by sashko** [comment]
I propose giving header files `.hpp` extension.

File: src/daemon/dbus_service.h (hunk 3 above)
Comment (TSAnsari) [reply src/daemon/dbus_service.h:30]: I don't mind if you use the glib standard names own_name and unown_name. Just didn't liked mixing them, like own and release.

**Comment by martin-ejdestig** [comment]
> I propose giving header files `.hpp` extension.

This is not that commonly done, in my experience. So I would object to doing this.

(Never worked in a project that has done it. And just looking through a couple of C++ projects I have laying around locally, e.g. Chromium, googletest, Qt, the Alexa SDK:s, Glibmm, libsigc++, Android, LLVM, Clang... etc. They all use .h for C++ headers.)

**Comment by sashko** [comment]
> > I propose giving header files `.hpp` extension.
> 
> This is not that commonly done, in my experience. So I would object to doing this.
> 
> (Never worked in a project that has done it. And just looking through a couple of C++ projects I have laying around locally, e.g. Chromium, googletest, Qt, the Alexa SDK:s, Glibmm, libsigc++, Android, LLVM, Clang... etc. They all use .h for C++ headers.)

In other words, people are wrong, so let's repeat after them.

File: src/daemon/dbus_service.cpp (hunk 2 above)
Comment (martin-ejdestig) [reply src/daemon/dbus_service.cpp:39]: Reverted commits that changed name. Resolving.

File: src/daemon/dbus_service.h (hunk 3 above)
Comment (martin-ejdestig) [reply src/daemon/dbus_service.h:30]: Reverted commits that changed name. Resolving.

**Comment by martin-ejdestig** [comment]
> In other words, people are wrong, so let's repeat after them.

Short reply. Yes. :)

More serious reply. I do not see (and have never seen) a problem with having them named `.h`. And without having any real data to back it up, I think it is what is most commonly done.

File: README.md (hunk 4 above)
Comment (martin-ejdestig) [reply README.md:4]: Removed extra \n. Resolving.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,12 @@
...
+
+# License and Copyright
+
+Copyright (C) 2019 Luxoft Sweden AB
```
Comment (sashko) [inline README.md]: &copy;

File: src/daemon/dbus_service.h [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+    private:
+        void bus_acquired(const Glib::RefPtr<Gio::DBus::Connection> &connection,
+                          const Glib::ustring &name);
+        void name_acquired(const Glib::RefPtr<Gio::DBus::Connection> &connection,
+                           const Glib::ustring &name);
+        void name_lost(const Glib::RefPtr<Gio::DBus::Connection> &connection,
+                       const Glib::ustring &name);
+
+        void RemoveMeFoo(guint32 number, MethodInvocation &invocation) override;
```
Comment (sashko) [inline src/daemon/dbus_service.h:41]: sO_wIcH_cOnVeNtIoN_aRe_We_UsIng()?

Comment (TSAnsari) [reply src/daemon/dbus_service.h:41]: I don't mind any convention as long as we are consistent and don't mix different styles. Perhaps this should be decided with the team, as this is a matter of personal preferences for many.

Comment (martin-ejdestig) [reply src/daemon/dbus_service.h:41]: This comes from the names in the D-Bus interface file and results in a method that must be overridden.

We can not be consistent with the outside world since it can contradict. E.g. two libraries can require methods to be overridden and they can have different coding styles. The `clang-tidy` `readability-identifier-naming` check understands this and does not enforce the specified naming for methods that are marked with `override`.

Resolving this. If you want to change style we use, bring it up independently . It is like it is here for a reason.

File: README.md (hunk 5 above)
Comment (martin-ejdestig) [reply README.md]: We seem to mix in our repos (and both are OK as far as I know). But OK, will change.

File: data/com.luxoft.TemplateDBusService.xml [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+      <arg name="number" type="u" direction="in"/>
+      <arg name="is_odd" type="b" direction="out"/>
+    </method>
+
+    <signal name="RemoveMeBar">
+      <arg name="something" type="u"/>
+    </signal>
+
+    <property name="RemoveMeBaz" type="b" access="readwrite"/>
+  </interface>
```
Comment (thenor) [inline data/com.luxoft.TemplateDBusService.xml:16]: Is it possible to have clang-format rules for xml files also? These interface/signal/property names are part of the discussion regarding snake/camel case.

File: src/daemon/dbus_service.h (hunk 6 above)
Comment (martin-ejdestig) [reply src/daemon/dbus_service.h:41]: If you want to change this in this particular case, I guess you can send a patch to `gdbus-codegen-glibmm` that makes it possible to choose style (and try to add _ where appropriate to camel case names... ugh!) or make it understand something like `org.gtk.GDBus.C.Name` (see https://developer.gnome.org/gio/stable/gdbus-codegen.html ). But if you add something like  `org.gtk.GDBus.C.Name`, you would enforce your style on users of the D-Bus introspection XML. A bit weird.

File: data/com.luxoft.TemplateDBusService.xml (hunk 7 above)
Comment (martin-ejdestig) [reply data/com.luxoft.TemplateDBusService.xml:16]: This is how methods should be named in D-Bus. See the D-Bus documentation.

By "clang-format" here I assume you mean if there is a tool to lint D-Bus introspection XML, not that `clang-format` itself should be used. I do not know of any such tool. The DTD is there so I assume some tool can be used to verify that, but it would not help with verifying naming convention used in the value for the `name`attribute.

File: meson.build (hunk 1 above)
Comment (martin-ejdestig) [reply meson.build:5]: Searching on https://layers.openembedded.org/layerindex/branch/master/layers/ and looking in the Poky repo with `gitk --all -- meta/recipes-devtools/gcc/` it looks to me like GCC 7.X (first major release May 2, 2017) is available for `thud`, `sumo` and even `pyro` but not `morty`.

**Comment by thenor** [comment]
I'm guessing there will be a squash of the commits. The commit messages aren't exactly descriptive.

File: data/com.luxoft.TemplateDBusService.xml (hunk 7 above)
Comment (thenor) [reply data/com.luxoft.TemplateDBusService.xml:16]: Oh it was the clang-tidy which contained the name formatting rules.
I was sort of assuming that there would be some naming convention already in place for the D-Bus methods.

**Comment by martin-ejdestig** [comment]
> I'm guessing there will be a squash of the commits.

Yes, have not rebased and force pushed due to GitHub's review system being bad. Force pushing will make original commits go away when they are garbage collected by GitHub (`git gc`), making it hard to follow old conversations etc. (At least if I have understood things correctly.)

`fixup!` (and `squash!`) commits is a `git` feature. Read about them in `git help commit` and `git help rebase` (perhaps there is better documentation out there, don't know). Commits will be squashed into commit with short message following `fixup!` when invoking `git rebase -i --autosquash`.

In this case it does not matter much. But it is really nice when working on a branch with multiple commits. (Well, that and it is easier to see what has actually been done to resolve "issues" in GitHub's b0rken review system.)

> The commit messages aren't exactly descriptive.

They are (assuming you only looked at the short message). Look at long description. The short description needs to be the way it is for the "autosquash" feature in git to work.

File: data/com.luxoft.TemplateDBusService.xml (hunk 7 above)
Comment (martin-ejdestig) [reply data/com.luxoft.TemplateDBusService.xml:16]: clang-tidy only checks C/C++ code, it knows nothing about D-Bus introspection XML. Probably not what you mean... but, hum... not sure I understand what is confusing here. To be extra clear.

- Convention in D-Bus is to name methods in CamelCase.
- gdbus-codegen-glibmm just takes the D-Bus method name in the introspection XML and generates a pure virtual method with that exact name in a C++ class (the stub).
- We subclass the stub and implement this pure virtual method when implementing a service.
- clang-tidy recognizes `override` and does not enforce the naming rules specified since... well... the method must be overridden... and comes from an external source (the code generated by the generator).

If we would have a hand written C++ class with a pure virtual method that did not follow the specified style, clang-tidy would have warned where that pure virtual method is specified, not where it is overridden.

File: meson.build (hunk 1 above)
Comment (JEderonn) [reply meson.build:5]: Ok, sounds good. Then we know we can build it with PELUX 1.0 since it's based on pyro, even though morty also should be supported in 1.0.
Let's use C++17 in this template.

File: data/com.luxoft.TemplateDBusService.xml (hunk 7 above)
Comment (thenor) [reply data/com.luxoft.TemplateDBusService.xml:16]: I meant that these same names are used as method names in dbus_service.cpp and there it makes it look odd, since it means not all method names use the same form.

Comment (martin-ejdestig) [reply data/com.luxoft.TemplateDBusService.xml:16]: Yes, but the name comes from an external source. Nothing to do about it. Just to drive my point home... e.g. say that we want to, in a single project:

- Subclass a QWidget and override https://doc.qt.io/qt-5/qwidget.html#heightForWidth .
- Subclass a Glib::IOChannel and override http://manual.freeshell.org/glibmm-2.4/reference/html/classGlib_1_1IOChannel.html#abfb4faa07e80b86f21b0f18f0b8606cd
- Use gdbus-codegen-glibmm, as in this project, and implement a D-Bus service where D-Bus specifies that method names are CamelCase and the generator just extracts the name from the XML and does not modify it in any way (as is done now).

Let's just say we do this in a single class for making the example simple. Would look something like this:

```C++
class Foo : <inherit from the stuff in the list above>
{
....
    // From QWidget. Uses lowerCaseCamelCase() style for methods.
    void hideEvent(QHideEvent *event) override;
....
    // From Glibmm. Uses snake_case() style for methods.
    Glib::IOStatus read_vfunc(char *buf, gsize count, gsize &bytes_read) override; 
...
    // From D-Bus. Uses CamelCase() style for methods.
    void RemoveMeFoo(MethodInvocation &invocation) override;
}
```

Three different styles, enforced on you from the outside world. It does not matter what style you prefer or enforce in other places of your project. These methods must be named the way they are since you `override`. You can not be consistent with all three of them at the same time.

And as I mentioned before, `clang-tidy` is smart here and does not warn about the identifier being wrongly named because it recognizes `override`.

Comment (thenor) [reply data/com.luxoft.TemplateDBusService.xml:16]: > Yes, but the name comes from an external source. Nothing to do about it. Just to drive my point home... e.g. say that we want to, in a single project:
> 

Then I am satisfied and know the reason for this difference in style. :)

> 
> And as I mentioned before, `clang-tidy` is smart here and does not warn about the identifier being wrongly named because it recognizes `override`.

Gottcha! :+1:

**Comment by martin-ejdestig** [comment]
As discussed on standup, it's not just good, it's good enough to merge at this point. :)

**Comment by martin-ejdestig** [comment]
Crusty's seal of approval
